Improve current GitHub Action CI with multi OS support

dyt811 opened the pull request:
<!-- Hi, and thank you for submitting a Pull Request! The checklist below is a brief summary of steps found in the NeuroPoly Contributing Guidelines, which can be found here: https://www.neuro.polymtl.ca/software/contributing. 
-->

## Checklist

#### GitHub

- [X] I've given this PR a concise, self-descriptive, and meaningful title
- [X] I've linked relevant issues in the PR body
- [X] I've applied [the relevant labels](https://www.neuro.polymtl.ca/software/contributing#pr_labels) to this PR
- [x] I've assigned a reviewer

<!-- For the title, please observe the following rules:
	- Provide a concise and self-descriptive title
	- Do not include the applicable issue number in the title, do it in the PR body
	- If the PR is not ready for review, convert it to a draft.
-->

#### PR contents

- [X] I've consulted [ivadomed's internal developer documentation](https://github.com/ivadomed/ivadomed/wiki) to ensure my contribution is in line with any relevant design decisions
- [X] I've added [relevant tests](https://github.com/ivadomed/ivadomed/wiki/tests) for my contribution
- [x] I've updated the [relevant documentation](https://github.com/ivadomed/ivadomed/wiki/documentation) for my changes, including argparse descriptions, docstrings, and ReadTheDocs tutorial pages

## Description
<!-- describe what the PR is about. Explain the approach and possible drawbacks. It's ok to repeat some text from the related issue. -->

### Why: 
- We currently have no CI for macOS/Windows
- Pip install -r requirement.txt will fail on Windows side because pytorch 1.5.0 does not have wheel on Windows
- We don’t have a Conda script to deploy environment even if we want to. 
- Dev don’t have quick way to deploy Dev environment (that includes pytest etc) via pip or conda 
- User/Dev don’t have quick way to deploy GPU related environment easily. 

### What:

- Current GitHub Action CI does not fully address Windows OR even macOS platform testing due to earlier bug which has been recently resolved. See #604 #755. Now we can properly robustly test Windows integration. I have incorporated many ADS CI tips and tricks learned from there (https://github.com/neuropoly/axondeepseg/pull/479) to improve our existing matrix testing to be more comprehensive and thorough while adding additional comments to make it easier for subsequent future updates. 

### How: 
Applying cascading `requirements.txt` without changing anything major it now allows us to accommodate various deployment scenarios while maintaining single source of truth
- `Requirements_common.txt` shared with `Conda` and `setup.py` cascade to various different higher level requirements. 
- `Conda` now supported via `environment.yml` which refer to the `requirements.txt` to respect SINGLE source of truth
- User deploys `requirements.txt` with CPU support, standard deployment scenario. 
- Advanced user deploys `requirements_gpu.txt` for GPU support. 
- Dev/CI/RTD uses `requirements_dev.txt` which cascades off the main requirements.txt

### How to test 
- Create new venv/conda env and validate using whichever conda/venv/pip installation setup approach you are most comfortable playing with. Report back issues and concerns. 
- `requirements_dev.txt` might require some extra attention as not all sphinx build/wheel are available for Windows? 
- Could also use more feedback on the documentation side

## Linked issues
<!-- If the PR fixes any issues, indicate it here with issue-closing keywords: e.g. Resolves #XX, Fixes #XX, Addresses #XX. Note that if you want multiple issues to be autoclosed on PR merge, you must use the issue-closing verb before each relevant issue: e.g. Resolves #1, Resolves #2 -->

Resolves #778, also provide CI for its precursor #633 #768

dyt811 committed:
WIP Significantly revised run_tests to support multi OS CI and Python 3.9

- Add numerous comments and lessons learned from ADS CI work
- Proper indentation upgrade (hence why so much red)

dyt811 committed:
Add additional step to use official py-dependency install instead

dyt811 committed:
Add force Python3 call to ensure consistent GitHub Python verison

dyt811 committed:
Reverse Python3 reference, remove bash -l parameter

dyt811 committed:
Update requirement to support 3.9 and pip install of torch CPU

dyt811 committed:
Ignore local default venv

dyt811 committed:
Separate requirements into four components that allow role specific installation

dyt811 committed:
Add conda support for users that wants it

dyt811 committed:
Update setup.py to use the common installation basis

dyt811 committed:
WIP: Update installation on latest three pronged installation approach 

- Still missing Docker installation

dyt811 committed:
WIP Add preliminary Docker basis

dyt811 committed:
Merge remote-tracking branch 'origin/master' into yd/740-windows_and_python3.9_CI

dyt811 committed:
Remove 3.9 from current PR consideration

dyt811 committed:
Update precommit to include dev related packages like pre-commit

dyt811 committed:
Updating Dev build to be based on the CPU requirement

dyt811 committed:
Merge branch 'master' into yd/740-windows_and_python3.9_CI

dyt811 committed:
Merge remote-tracking branch 'origin/master' into yd/740-windows_and_python3.9_CI

dyt811 committed:
Add precedent requirements.txt installation for PyTorch1.5.0

Adding requirements before setup.py to ensure the PyTorch is included in the environment.

dyt811 committed:
Correct formatting for requirements.txt

- This should allow readthedoc to pass hopefully.

dyt811 committed:
Addressing missing dev dependencies after requirements.txt file refactoring

dyt811 commented:
Took a while to realize that macOS does not support the Wheel +CPU tag... sigh. 
![image](https://user-images.githubusercontent.com/6720068/117579653-e90c3480-b0c1-11eb-9893-62e28243aa14.png)

No wonder it fails on macOS only. FML.

dyt811 committed:
Environment update

- Conda env now use CPU from pip by default
- Requirements now does os specific handling of Torch installation
- Streamlined GPU requirement installation process

dyt811 committed:
Merge remote-tracking branch 'origin/master' into yd/740-windows_and_python3.9_CI

dyt811 committed:
Attempt to address RTD failure

- RTD seems skipped torch build. Could it be using JAVA?

dyt811 committed:
Corrected the env variable checked for pip requirements

dyt811 committed:
Remove docker related changes to next PR

dyt811 committed:
Incorporate additional specific OS version into CI

dyt811 committed:
Update documentation per new Conda vs Venv/PIP Approach

dyt811 committed:
Fix YAML missing OS key

dyt811 committed:
Disabled interactive plot to attempt avoid CI stall

dyt811 committed:
Attempt Yuan's solution to resolve CI locking

charleygros commented:
@dyt811 : Thanks for all this! I'd be happy to review, but I'm scared I don't have a relevant skill set for this. Maybe someone else in the team would be more appropriate? Let me know :-)

dyt811 committed:
Add missing script_runner fixture import causing CI skips

ssinad committed:
Add multi-OS support to the CI tests (#785)

* Change Ubuntu to Windows

* Change Ubuntu to Windows

* Add PowerShell If

* Add CPU version command for Windows

* Windows is not needed for coveralls

* Use requirements.txt

* Rename test

* Add both operating systems

* Add both operating systems

* Fix test bug

* Add Python 3.9

* Testing again

* Testing again

* Update torch version in requirements.txt

* Loosen dependencies

* Update dependencies

* Debugging

* Debugging

* Testing

* Testing

* Testing

* Add -f option back

* Change requirements

* Add macos to the CI

* macos 11.0 is not supported yet

* Add DEBIAN_FRONTEND variable

* Unify scripts

* Add both versions of ubuntu

* Fix bug

* Remove ubuntu-latest from the matrix to reduce ambiguity

* Add multi-OS support to pre-commit script

* Fix pre-commit.yml bug

* Set shell to bash for all Operating systems

* Add python-version to matrix in pre-commit.yml

dyt811 committed:
Add missing pytest-console-scripts required for Automated Testing

dyt811 committed:
Merge remote-tracking branch 'origin/yd/740-windows_and_python3.9_CI' into yd/740-windows_and_python3.9_CI

dyt811 commented:
Closer examination of the output showed that automated_training_process was skipped accidentally. Once properly restored, it suggested that we are still running into issue with Mac/Windows hanging and it is on top of the earlier mentioned problem with HDF5. Root cause still not fully identified...

dyt811 committed:
Made ViolinPlot optional to prevent CI blocking

- Past local unit tests

dyt811 committed:
Merge remote-tracking branch 'origin/master' into yd/740-windows_and_python3.9_CI

dyt811 committed:
Undo ineffective earlier attempt to unblock CI

-  These lines were introduced to attempt unblock CI. Was not the cause. Reverting to original.

ssinad commented on .github/workflows/run_tests.yml:
```diff
@@ -1,62 +1,100 @@
+
+      # Step 2: Install requirement dependencies.
+      - name: Install dependencies
+        uses: py-actions/py-dependency-install@v2
```
@dyt811 Which requirements file does this one install?

ssinad replied:
Maybe you could add
```yaml
path: requirements_dev.txt
```

ssinad commented on environment.yml:
```diff
@@ -0,0 +1,10 @@
+name: IvadoMedEnv
```
@dyt811  Could you explain what the purpose of this file is?

dyt811 replied:
This environment YML is provided such that conda users can easily instantiate a conda environment while still be able to source the configure file from a single source of truth which is shared across multiple dev/user environment setups such as: 1) CPU Dev, 2) GPU users. 

Conda user can certainly just instantiate their own environment and pip install the requirement as well, just save them a few key strokes.

ssinad commented on setup.py:
```diff
@@ -2,7 +2,9 @@
+# Manually specified, more generic version of the software.
+# See: https://stackoverflow.com/a/49684835
+with open('requirements_common.txt') as f:
     requirements = f.read().splitlines()
```
```python
requirements = f.readlines()
```
would also work.

dyt811 replied:
TIL. Cool. Definitely didn't know that before. :)

ssinad reviewed (changes requested):
@dyt811
I would recommend that you get another approval from someone else on the team before you merge.

cakester commented on requirements_dev.txt:
```diff
@@ -0,0 +1,13 @@
+sphinx
+pypandoc
+sphinx_rtd_theme
+sphinx-jsonschema
```
no version number? in the context that you just deleted this has a version number https://github.com/ivadomed/ivadomed/pull/757/files#diff-4d7c51b1efe9043e44439a949dfd92e5827321b34082903477fd04876edb7552L19

dyt811 replied:
Good catch. Yeah, I think this PR started before your PR so this needs to be manually added in this new file. I will added it.

ssinad commented on .github/workflows/run_tests.yml:
```diff
@@ -1,62 +1,100 @@
+      - name: Install dev dependencies
+        run: |
+          pip install flake8 pytest coverage coveralls pytest-cov pytest-console-scripts pytest-ordering
+          pip install torch==1.5.0+cpu torchvision==0.6.0+cpu -f https://download.pytorch.org/whl/torch_stable.html
```
@dyt811 Consider using the previous step instead of hardcoding package versions.

dyt811 replied:
Good point!

jcohenadad commented on docs/source/installation.rst:
```diff
@@ -27,8 +27,40 @@ Write in your .bashrc/.zshrc file with following line.
 
-Python
-------
+There are fundamentally two different approaches to install IvadoMed:
```
```suggestion
There are fundamentally two different approaches to install ``ivadomed``:
```

ssinad commented on .github/workflows/run_tests.yml:
```diff
@@ -1,62 +1,100 @@
+
+      # Step 2: Install requirement dependencies.
+      - name: Install dependencies
+        uses: py-actions/py-dependency-install@v2
```
```suggestion
        uses: py-actions/py-dependency-install@v2
        path: requirements_dev.txt
```

dyt811 replied:
Yup. Change included.

ssinad commented on setup.py:
```diff
@@ -2,7 +2,9 @@
+# Manually specified, more generic version of the software.
+# See: https://stackoverflow.com/a/49684835
+with open('requirements_common.txt') as f:
     requirements = f.read().splitlines()
```
```suggestion
    requirements = f.readlines()
```

dyt811 replied:
Updated.

ssinad commented on .github/workflows/pre-commit.yml:
```diff
@@ -27,13 +36,17 @@ jobs:
         run: |
               python -m pip install --upgrade pip
-              if [ -f requirements.txt ]; then pip install -r requirements.txt; fi
+              if [ -f requirements_dev.txt ]; then pip install -r requirements_dev.txt; fi
```
```suggestion
              if [[ -f requirements_dev.txt ]]; then pip install -r requirements_dev.txt; fi
```
See [here](https://github.com/ivadomed/ivadomed/wiki/variables#if-statements).

dyt811 replied:
COOL. TIL. Thanks for the info!

dyt811 committed:
Merge remote-tracking branch 'origin/master' into yd/740-windows_and_python3.9_CI

dyt811 commented on .github/workflows/run_tests.yml:
```diff
@@ -1,62 +1,100 @@
+    # Default shell for ALL subsequent steps.
+    defaults:
+      run:
+        shell: bash {0}
```
I believe this [disables fast fail in GitHub CI](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#exit-codes-and-error-action-preference) which result in more likelihood of revealing bugs in CI. I have actually [copied this bit from the template GitHub Action from ADS](https://github.com/neuropoly/axondeepseg/blob/master/.github/workflows/run_tests.yaml#L44).

dyt811 committed:
Incorporating all dev feedbacks

- Minor stylistic improve in documentation, bash if [[]]
- CI step now use the requirement_dev.txt
- Add versioning for requirements.txt

ssinad reviewed (approved):
Great job. I would still recommend you get another approval before you merge, though.

jcohenadad commented:
> Great job. I would still recommend you get another approval before you merge, though.

we are a very small team and there are *many* other urgent priorities so unfortunately we need to live with PRs not being reviewed as much as everyone would like--

ssinad commented:
> > Great job. I would still recommend you get another approval before you merge, though.
> 
> we are a very small team and there are _many_ other urgent priorities so unfortunately we need to live with PRs not being reviewed as much as everyone would like--

I understand.

ssinad commented on requirements_dev.txt:
```diff
@@ -0,0 +1,18 @@
+
+# Dev:
+# Development, documentation, testing related dependencies.
+pytest
```
```suggestion
pytest~=6.2
```

dyt811 replied:
Good catch. Yup. Better to version these. Gonna add it in.

ssinad commented on requirements_dev.txt:
```diff
@@ -0,0 +1,18 @@
+pytest
+pytest-cov
+pytest-console-scripts
+pytest-ordering
```
```suggestion
pytest-ordering~=0.6
```

dyt811 replied:
Yup. Will add this in as well. I was REALLY hoping that we will get CI to a state where daily CI runs will catch any subversion induced problems but I think there is a whole bunch of tests to enhance before that happens.

dyt811 committed:
Incorporated additional Dev dependency lockdowns

dyt811 committed:
Add BigSur to the macOS testings

konantian commented on .github/workflows/pre-commit.yml:
```diff
@@ -8,15 +8,24 @@ on:
-    runs-on: 'ubuntu-latest'
+    strategy:
+      matrix:
+        os: [ "macos-11", "windows-latest", "ubuntu-18.04", "ubuntu-20.04",  "macos-10.15"]
```
We use "macos-11" here is because "macos-latest" is equals to "macos-10.15" instead of "macos-11".

konantian commented on .github/workflows/run_tests.yml:
```diff
@@ -1,62 +1,100 @@
+
       matrix:
-        python-version: [3.6, 3.7, 3.8]
+        os: [ "macos-11", "windows-latest", "ubuntu-18.04", "ubuntu-20.04",  "macos-10.15" ]
```
We use "macos-11" here is because "macos-latest" is equals to "macos-10.15" instead of "macos-11".

ssinad replied:
@dyt811 
macOS 11 is not available in GitHub actions yet. That is why the tests are failing. See [here](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners).

> Note: The macOS 11 virtual environment is currently provided as a private preview only. Any users or organizations that are already using this runner can continue using it, but we're not accepting any further users or organizations at this time. The macos-latest YAML workflow label still uses the macOS 10.15 virtual environment.

dyt811 replied:
Good point. I thought it was available but didn't read the fine print about it being "private preview only". Gonna revert that and merge PR then.

konantian reviewed (commented):
Looks good to me, I've tested them on both Windows and macOs locally, no issues found.

dyt811 committed:
Revert macOS11 GitHub specification

- Using macos-latest when macos11 becomes available in the future

dyt811 committed:
Merge remote-tracking branch 'origin/master' into yd/740-windows_and_python3.9_CI

dyt811 committed:
Merge remote-tracking branch 'origin/master' into yd/740-windows_and_python3.9_CI

dyt811 committed:
Correct typo in dependency name

kousu commented on environment.yml:
```diff
@@ -0,0 +1,10 @@
+  - pytorch
+dependencies:
+  - python>=3.6, < 3.9
+  - anaconda
```
@dyt811 can you explain why you depend on `anaconda` explicitly?

This doesn't seem to be compatible with https://github.com/conda-forge/miniforge, which is what I picked to use in https://github.com/neuropoly/computers/pull/120/ on the theory that it would be smaller and better supported.

Removing this one line seems to make ivadomed fully compatible with miniforge.

dyt811 replied:
Yup. No need to use anaconda. Miniconda is good enough. 👍
